20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,558 @@
+
+When the time comes, check will be updated to ignore relationship tuples in the database that do not match the directly related user types in the authorization model.
+
+## Migration
+
+### Migrating existing authorization models
+
+Existing models will not be migrated - new models will be required to use types (with an optional grace period). Authorization models that do not make use of types will not be able to use the optimized ListObjects endpoint and will fallback to the existing brute force implementation.
+
+### Migrating existing tuples
+
+Due to the users in the tuples now required to have types in order to enforce the restrictions, existing tuples with users as floating user identifiers with no types will no longer be valid when types are added. These tuples will not be removed from the system, and will still be valid while the legacy authorization model is supported, but will be ignored during evaluation on newer authorization models with type restrictions in place.
+
+This will need to be communicated to developers so that they can migrate accordingly by:
+
+1. Introducing a `user` type to the model
+2. Reading all exiting relationship tuples to find ones with a user that has no type
+3. Writing a mirror of that tuple where the user has the correct type
+4. Migrating their app code to perform checks using that type
+
+One option for developers administrating an OpenFGA installation could be a script to check the DB (look for tuples in the DB that do not have `user_type`), and prints the offending store IDs.
+
+Automatic unassisted tuple migration will not be feasible because it is not possible to know what type end users will want to use for each offending tuple.
+
+### Affected Modules
+
+This change will affect repositories across the board. It will entail changes to the [protobuf](https://github.com/openfga/api), the [openfga core](https://github.com/openfga/openfga), the DSL, the [syntax transformer](https://github.com/openfga/syntax-transformer), the [SDKs](https://github.com/openfga/sdk-generator), the FGA Playground, the [sample stores](https://github.com/openfga/sample-stores) and the [documentation](https://github.com/openfga/openfga.dev).
+
+#### Language, API and SDKs
+
+For the scope of this RFC, we are proposing that the initial phase be backwards compatible and not a breaking change. This will allow users on previous versions to keep using them for a while, similarly changes to the public surface of the API and SDKs are reduced.
+
+A later RFC can introduce the upgraded syntax where the extension now placed in metadata can be migrated. That phase can be combined with other breaking changes we are introducing to minimize user-disruption.
+
+#### DSL
+
+An update to the DSL needs to happen to support type with no definitions, as a lot of developers will now have to include a user type and it may not have any relations on it.
```

Comment:
```suggestion
As a lot of developers will now have to include a user type and it may not have any relations on it, an update to the DSL needs to happen to support types with no relations:

`type user
  relations none
`
```